File: Android/doraemonkit/src/main/java/com/didichuxing/doraemonkit/aop/bigimg/glide/DokitGlideTransform.java
package com.didichuxing.doraemonkit.aop.bigimg.glide;

import android.graphics.Bitmap;

import androidx.annotation.NonNull;

import com.blankj.utilcode.util.ReflectUtils;
import com.bumptech.glide.RequestBuilder;
import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool;
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation;
import com.didichuxing.doraemonkit.kit.largepicture.LargePictureManager;
import com.didichuxing.doraemonkit.util.LogHelper;

import java.security.MessageDigest;

/**
 * ================================================
 * 作    者：jint（金台）
 * 版    本：1.0
 * 创建日期：2020/4/3-13:04
 * 描    述：
 * 修订历史：
 * ================================================
 */
public class DokitGlideTransform extends BitmapTransformation {
    private static final String ID = "com.didichuxing.doraemonkit.aop.bigimg.glide.DokitGlideTransform";
    private static final byte[] ID_BYTES = ID.getBytes(CHARSET);
    private static final String TAG = "DokitGlideTransform";
    private Object mRequestBuilder;

    public DokitGlideTransform(Object mRequestBuilder) {
        this.mRequestBuilder = mRequestBuilder;
    }

    @Override
    protected Bitmap transform(@NonNull BitmapPool pool, @NonNull Bitmap toTransform, int outWidth, int outHeight) {
        LogHelper.i(TAG, "===transform=====");
        String url = "";
        if (mRequestBuilder instanceof RequestBuilder) {
            url = ReflectUtils.reflect(mRequestBuilder).field("model").get();
        }
        toTransform = LargePictureManager.getInstance().transform(url, toTransform, false, "Glide");
        return toTransform;
    }

    @Override
    public void updateDiskCacheKey(@NonNull MessageDigest messageDigest) {
        messageDigest.update(ID_BYTES);
    }
}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/aop/bigimg/glide/DokitGlideTransform.kt
package com.didichuxing.doraemonkit.aop.bigimg.glide

import android.content.Context
import android.graphics.Bitmap
import com.bumptech.glide.load.Key
import com.bumptech.glide.load.Transformation
import com.bumptech.glide.load.engine.Resource
import java.security.MessageDigest

/**
 * ================================================
 * 作    者：jint（金台）
 * 版    本：1.0
 * 创建日期：2020/4/3-13:04
 * 描    述：
 * 修订历史：
 * ================================================
 */
class DokitGlideTransform(private val mRequestBuilder: Any, transformation: Any?) : Transformation<Bitmap?> {
    private var mWrap: Transformation<Bitmap>? = null
    override fun updateDiskCacheKey(messageDigest: MessageDigest) {
        if (mWrap != null) {
            mWrap?.updateDiskCacheKey(messageDigest)
        } else {
            messageDigest.update(ID_BYTES)
        }
    }

    override fun equals(o: Any?): Boolean {
        return if (mWrap != null) {
            mWrap == o
        } else false
    }

    override fun hashCode(): Int {
        return mWrap?.hashCode() ?: 0
    }

    override fun transform(context: Context, resource: Resource<Bitmap?>, outWidth: Int, outHeight: Int): Resource<Bitmap?> {
        //TODO("功能待实现")
//        var resource = resource
//        try {
//            if (mWrap != null) {
//                resource = mWrap.transform(context, resource, outWidth, outHeight)
//            }
//            if (PerformanceSpInfoConfig.isLargeImgOpen()) {
//                var url: String? = ""
//                if (mRequestBuilder is RequestBuilder<*>) {
//                    if (ReflectUtils.reflect(mRequestBuilder).field("model").get<Any>() is String) {
//                        url = ReflectUtils.reflect(mRequestBuilder).field("model").get()
//                    } else if (ReflectUtils.reflect(mRequestBuilder).field("model").get<Any>() is Int) {
//                        url = "" + ReflectUtils.reflect(mRequestBuilder).field("model").get()
//                    }
//                }
//                val bitmap = resource.get()
//                val imgSize = ConvertUtils.byte2MemorySize(bitmap.byteCount.toLong(), MemoryConstants.MB)
//                LargePictureManager.getInstance().saveImageInfo(url, imgSize, bitmap.width, bitmap.height, "Glide")
//            }
//        } catch (e: Exception) {
//            if (mWrap != null) {
//                resource = mWrap.transform(context, resource, outWidth, outHeight)
//            }
//        }
        return resource
    }

    companion object {
        private const val ID = "com.didichuxing.doraemonkit.aop.bigimg.glide.DokitGlideTransform"
        private val ID_BYTES = ID.toByteArray(Key.CHARSET)
        private const val TAG = "DokitGlideTransform"
    }

    init {
        if (transformation is Transformation<*>) {
            mWrap = transformation as Transformation<Bitmap>
        }
    }
}